cmd/env.go

```diff
@@ -53,24 +86,113 @@ func env(cmd *cobra.Command, args []string) error {
 		})
 	}
 
+	params := make(map[string]string)
 	for _, rawSecret := range rawSecrets {
-		fmt.Printf("export %s=%s\n",
-			strings.ToUpper(key(rawSecret.Key)),
-			shellescape(rawSecret.Value))
+		params[key(rawSecret.Key)] = rawSecret.Value
 	}
 
-	return nil
+	out, err := buildEnvOutput(params)
+	if err != nil {
+		return nil, err
+	}
+
+	// ensure output prints variable declarations as exported
+	for i := range out {
+		// Sprintf because each declaration already ends in a newline
+		out[i] = fmt.Sprintf("export %s", out[i])
+	}
+
+	return out, nil
 }
 
-// shellescape returns a shell-escaped version of the string s. The returned value
-// is a string that can safely be used as one token in a shell command line.
-func shellescape(s string) string {
-	if len(s) == 0 {
-		return "''"
+// output will be returned lexically sorted by key name
+func buildEnvOutput(params map[string]string) ([]string, error) {
+	out := []string{}
+	for _, key := range sortedKeys(params) {
+		name := sanitizeKey(key)
+		if !preserveCase {
+			name = strings.ToUpper(name)
+		}
+
+		if err := validateShellName(name); err != nil {
+			return nil, err
+		}
+
+		// the default format prints all escape sequences as
+		// string literals, and wraps values in single quotes
+		// if they're unsafe or multi-line strings.
+		s := fmt.Sprintf(`%s=%s`, name, shellescape.Quote(params[key]))
+		if escapeSpecials {
+			// this format collapses special characters like newlines
+			// or carriage returns. requires escape sequences to be interpolated
+			// by whatever parses our key="value" pairs.
+			s = fmt.Sprintf(`%s="%s"`, name, doubleQuoteEscape(params[key]))
+		}
+
+		// don't rely on printf to handle properly quoting or
+		// escaping shell output -- just white-knuckle it ourselves.
+
+		out = append(out, s)
 	}
-	if pattern.MatchString(s) {
-		return "'" + strings.Replace(s, "'", "'\"'\"'", -1) + "'"
+
+	return out, nil
+}
+
+// The name of a variable can contain only letters (a-z, case insensitive),
+// numbers (0-9) or the underscore character (_). It may only begin with
+// a letter or an underscore.
+func validateShellName(s string) error {
+	shellChars := regexp.MustCompile(`^[A-Za-z0-9_]+$`).MatchString
+	validShellName := regexp.MustCompile(`^[A-Za-z_]{1}`).MatchString
```

Comment:
for these regexps, would it make sense to add them to our [list of valid format regexps](https://github.com/segmentio/chamber/blob/e032a23a8c16f4bd86f0da03bdbf2127dfc89717/cmd/root.go#L19-L23)?